### PR TITLE
SourceChecker reorganization

### PR DESCRIPTION
**Pull request by mernst**
This pull request moves methods around in the file to improve the file's organization.
It renames a few methods and fields.
It adds documentation.
It abstracts out two methods (to make the calling method, which used to inline the code shorter and clearer).

There are no logic changes in the pull request.

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java [hunk 1]
```diff
@@ -499,6 +504,31 @@ public final synchronized void init(ProcessingEnvironment env) {
         }
     }
 
+    /** True if the git.properties file has been printed. */
+    private static boolean gitPropertiesPrinted = false;
+
+    /** Print information about the git repository from which the Checker Framework was compiled. */
+    void printGitProperties() {
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java]: Should this method be private? Or protected? I don't see a need for package visibility.

File: framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java [hunk 2]
```diff
@@ -499,6 +504,31 @@ public final synchronized void init(ProcessingEnvironment env) {
         }
     }
 
+    /** True if the git.properties file has been printed. */
+    private static boolean gitPropertiesPrinted = false;
+
+    /** Print information about the git repository from which the Checker Framework was compiled. */
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java]: Could we move/keep these support/debugging methods to the bottom of the file? Is there a reason to move them up?

File: framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java [hunk 3]
```diff
@@ -1066,178 +895,79 @@ public void typeProcess(TypeElement e, TreePath p) {
...
-                    // For example, @SuppressWarnings("purity") suppresses errors with keys:
-                    // purity.deterministic.void.method, purity.deterministic.constructor, etc.
-                    if (errorKey.contains(userKey)) {
-                        reportUnneededSuppression(tree, fullUserKey);
-                    }
-                }
-            }
-        }
-    }
+    ///////////////////////////////////////////////////////////////////////////
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java]: I would also move shutdown logic towards the bottom.

File: framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java [hunk 4]
```diff
@@ -1066,178 +895,79 @@ public void typeProcess(TypeElement e, TreePath p) {
...
-                        + ")"
-                        + ((p == null)
-                                ? ""
-                                : " while processing "
-                                        + p.getCompilationUnit().getSourceFile().getName())
-                        + (t.getMessage() == null ? "" : "; message: " + t.getMessage()),
-                t);
-    }
+    ///////////////////////////////////////////////////////////////////////////
+    /// Reporting type-checking errors; the main entry point is report()
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java]: I'm not sure what the "main entry point" is supposed to mean. Can you clarify?
The two methods that visitors should use are `report()` and `message(Diagnostic.Kind kind, String msg, Object... args)`, so maybe list those two?

File: framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java [hunk 5]
```diff
@@ -2053,100 +1634,551 @@ public final String getOption(String name, String defaultValue) {
...
+    /**
+     * Issues a warning about any {@code @SuppressWarnings} that isn't used by this checker, but
+     * contains a key that would suppress a warning from this checker.
+     *
+     * @param elementsSuppress elements with a {@code @SuppressWarnings} that actually suppressed a
+     *     warning
+     * @param checkerKeys suppress warning keys that suppress any warning from this checker
+     * @param errorKeys error keys that can be issued by this checker
+     */
+    protected void warnUnneedSuppressions(
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java]: Should this method be called `warnUnneededSuppressions`, like the other method above?
How does a developer chose between the two version?

File: framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java (hunk 1 above)
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java]: I made it private.

File: framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java (hunk 2 above)
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java]: It is only called by `init`, so I put it next to `init`.  I have now moved it to the end of the file.

File: framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java (hunk 4 above)
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java]: `message` is documented as being "only for exceptional cases".  So `report` is the method that clients should prefer.

File: framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java (hunk 5 above)
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java]: That was the name before, but I have changed it.

**Pull request merged**